qiskit/circuit/parametertable.py

```diff
@@ -12,43 +12,140 @@
 """
 Look-up table for variable parameters in QuantumCircuit.
 """
-import warnings
 import functools
-from collections.abc import MutableMapping, MappingView
+import warnings
+from collections.abc import MappingView, MutableMapping, Set
+from typing import MutableSequence
+
+
+class ParameterReferences(MutableSequence, Set):
```

Comment:
Is there a reason to maintain the insertion order here other than just backwards compatibility?  The mutable set interface with a custom key seems to be much more natural for what we need.

Reply:
Currently, there are a number of places in Terra that depend on parameter references being indexable, so I assumed it was something we should preserve. I can take a look at those callsites again and see if I can rewrite the logic if you think that'd be better. Thoughts? :)

Reply:
Is there anywhere that _logically_ need to access by index?  Off the top of my head, I wouldn't have expected so - I can't think of why you'd need an ordering to them, and there's a good chance that the iterable form is causing performance issues in things like duplication checking (I know some of that happens in `QuantumCircuit` for sure, which is presumably the point of this PR).  If there's nothing that needs the sequence, it'd be good to drop it entirely, I think.  In that case, it would become `MutableSet`, not just `Set`.

Reply:
Agreed—it'd be great to just have this be a `MutableSet`. I'll look into it and report back. Originally, I took the shortcut of assuming there was a logical reason to index into the sequence since to me, it seemed like an odd thing to do otherwise :).